### PR TITLE
Px4 hil plugins

### PR DESCRIPTION
**Pull request by mzahana**
This branch adds support for MAVLink messages that are used in HIL with PX4 firmware.

Added messages:
- HilGPS, HilControls, HILSensor, HilStateQuaternion, ActuatorControlTarget

Plugins:
- hil_controls, hil_sensor, hil_state_quaternion, hil_gps, actuator_control_target

This is demonstrated with VREP simulator. [Demo video](https://www.youtube.com/watch?v=NoEw4upSE4M&feature=youtu.be)

### CODE REVIEW

File: mavros/include/mavros/hil_sensor_mixin.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+namespace mavros {
+namespace plugin {
+/**
+ * @brief This mixin adds set_hil_sensor()
+ */
+template <class D>
+class SetHilSensorMixin {
+public:
+	//! Message specification: @p https://pixhawk.ethz.ch/mavlink/#HIL_SENSOR
+	void set_hil_sensor(uint64_t time_boot_us,
```
Comment (vooon) [inline mavros/include/mavros/hil_sensor_mixin.h]: This function used only once! Mixing class absolutely not needed.

File: mavros/include/mavros/hil_sensor_mixin.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+};
+
+/**
+ * @brief This mixin adds set_hil_state_quaternion()
+ */
+template <class D>
+class SetHilStateQuaternionMixin {
+public:
+    //! Message specification: @p https://pixhawk.ethz.ch/mavlink/#HIL_STATE_QUATERNION
+    void set_hil_state_quaternion(uint64_t time_boot_us,
```
Comment (vooon) [inline mavros/include/mavros/hil_sensor_mixin.h]: Again. No reason for mixin class.

File: mavros/include/mavros/hil_sensor_mixin.h [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+};
+
+/**
+ * @brief This mixin adds set_hil_gps()
+ */
+template <class D>
+class SetHilGPSMixin {
+public:
+    //! Message specification: @p https://pixhawk.ethz.ch/mavlink/#HIL_GPS
+    void set_hil_gps(uint64_t time_boot_us, uint8_t fix_type,
```
Comment (vooon) [inline mavros/include/mavros/hil_sensor_mixin.h]: And again…

File: mavros/src/plugins/actuator_control_target.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#include <mavros/mavros_plugin.h>
+
+#include <mavros_msgs/ActuatorControlTarget.h>
+
+namespace mavros {
+namespace std_plugins {
+/**
+ * @brief Actuator Control Target plugin
+ */
+class ActuatorControlTargetPlugin : public plugin::PluginBase {
```
Comment (vooon) [inline mavros/src/plugins/actuator_control_target.cpp]: Any reason not to extend ActuatorControlPlugin?

File: mavros/src/plugins/hil_sensor.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+					temperature,
+					fields_updated);
+	}
+
+	/* -*- callbacks -*- */
+        
+
+		void sensor_cb(const mavros_msgs::HilSensor::ConstPtr &req) {
+            
+            send_hil_sensor(req->header.stamp,
```
Comment (vooon) [inline mavros/src/plugins/hil_sensor.cpp]: No reason for all that single-use functions (except callback off course).

File: mavros_msgs/msg/ActuatorControlTarget.msg [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# ActuatorControlTarget.msg
```
Comment (vooon) [inline mavros_msgs/msg/ActuatorControlTarget.msg]: Why you need it? I don't see any difference with ActuatorControl.msg.

File: mavros_msgs/msg/HilGPS.msg [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+uint8 fix_type
+int32 lat
+int32 lon
+int32 alt
+uint16 eph
+uint16 epv
+uint16 vel
+int16 vn
+int16 ve
+int16 vd
```
Comment (vooon) [inline mavros_msgs/msg/HilGPS.msg]: I'd prefer to use verctors, but you use unscaled values. Maybe ok.

File: mavros_msgs/msg/HilSensor.msg [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+std_msgs/Header header
+float32 xacc
+float32 yacc
+float32 zacc
+float32 xgyro
+float32 ygyro
+float32 zgyro
+float32 xmag
+float32 ymag
+float32 zmag
```
Comment (vooon) [inline mavros_msgs/msg/HilSensor.msg]: Here i would like to see vectors.

File: mavros_msgs/msg/HilStateQuaternion.msg [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# HilStateQuaternion.msg
+#
+# ROS representation of MAVLink HIL_STATE_QUATERNION
+# See mavlink message documentation here:
+# https://pixhawk.ethz.ch/mavlink/#HIL_STATE_QUATERNION
+
+std_msgs/Header header
+float32[4] quat_wxyz
```
Comment (vooon) [inline mavros_msgs/msg/HilStateQuaternion.msg]: geometry_msgs/Quaternion!

**Comment by v01d** [comment]
@mzahana I'm interested in these changes. Do you plan on finishing this PR?

**Comment by mzahana** [comment]
@v01d Yes, I plan to finish this PR. However, not any soon! I am currently supper busy even during my free time. I will have more free time after March. Sorry for inconvenience.

**Comment by TSC21** [comment]
Closing in favor of #684. @mzahana please check new PR and see if there's anything you would like to add.

**Pull request closed without merging**